src/codegen/encoding/mod.rs

```diff
@@ -10,27 +10,520 @@ use crate::codegen::encoding::scale_encoding::ScaleEncoding;
+            Type::Uint(width) | Type::Int(width) => {
+                self.encode_int(expr, buffer, offset, vartab, cfg, *width)
+            }
+            Type::Value => {
+                let size = ns.value_length.into();
+                self.encode_direct(expr, buffer, offset, vartab, cfg, size)
+            }
+            Type::Bytes(length) => {
+                self.encode_direct(expr, buffer, offset, vartab, cfg, (*length).into())
+            }
+            Type::String | Type::DynamicBytes => {
+                self.encode_bytes(expr, buffer, offset, vartab, cfg)
+            }
+            Type::Enum(_) => self.encode_direct(expr, buffer, offset, vartab, cfg, 1.into()),
+            Type::Struct(ty) => {
+                self.encode_struct(expr, buffer, offset.clone(), ty, arg_no, ns, vartab, cfg)
+            }
+            Type::Slice(ty) => {
+                let dims = &[ArrayLength::Dynamic];
+                self.encode_array(
+                    expr, expr_ty, ty, dims, arg_no, buffer, offset, ns, vartab, cfg,
+                )
+            }
+            Type::Array(ty, dims) => self.encode_array(
+                expr, expr_ty, ty, dims, arg_no, buffer, offset, ns, vartab, cfg,
+            ),
+            Type::ExternalFunction { .. } => {
+                self.encode_external_function(expr, buffer, offset, ns, vartab, cfg)
+            }
+            Type::FunctionSelector => {
+                let size = ns.target.selector_length().into();
+                self.encode_direct(expr, buffer, offset, vartab, cfg, size)
+            }
+            Type::Ref(r) => {
+                if let Type::Struct(ty) = &**r {
+                    // Structs references should not be dereferenced
+                    return self.encode_struct(
+                        expr,
+                        buffer,
+                        offset.clone(),
+                        ty,
+                        arg_no,
+                        ns,
+                        vartab,
+                        cfg,
+                    );
+                }
+                let loaded = Expression::Load(Codegen, *r.clone(), expr.clone().into());
+                self.encode(&loaded, buffer, offset, arg_no, ns, vartab, cfg)
+            }
+            Type::StorageRef(..) => {
+                let loaded = self.storage_cache_remove(arg_no).unwrap();
+                self.encode(&loaded, buffer, offset, arg_no, ns, vartab, cfg)
+            }
+            Type::UserType(_) | Type::Unresolved | Type::Rational | Type::Unreachable => {
+                unreachable!("Type should not exist in codegen")
+            }
+            Type::InternalFunction { .. }
+            | Type::Void
+            | Type::BufferPointer
+            | Type::Mapping(..) => unreachable!("This type cannot be encoded"),
+        }
+    }
+
+    /// Write whatever is inside the given `expr` into `buffer` without any modification.
+    fn encode_direct(
```

Comment:
This should be `encode_directly`.

Reply:
Ok. Why is this important?

Reply:
> Ok. Why is this important?

Your functions encodes the type directly in the buffer, so you need an adverb.

Reply:
I don't think this matter much. It does not improve the function name.

Reply:
> I don't think this matters much. It does not improve the function name.

It is proper grammar :)

Reply:
[I also don't think it matters much ](https://en.wikipedia.org/wiki/Law_of_triviality) :wink: Changed it, but I won't keep bikeshdding the PR after this.

Reply:
function naming is not prose, so strict grammar rules don't apply.